File: service/comments.ml
(*
 * file : comments.ml
 * author : KDr2
 *)

let opt_str_nn v = match v with
  | Some s -> s
  | None -> "";;

let comments_callback ret row headers =
  let comment = ref [] in begin
    for i = 0 to Array.length headers - 1 do
      comment := (headers.(i),Json_type.String (opt_str_nn row.(i)))::!comment
    done;
    match !ret with
      | Json_type.Array l ->
        ret := Json_type.Array ((Json_type.Object !comment) :: l)
      | _ -> ()
  end;;


let get_comments (cgi:Netcgi.cgi) arg_map () =
  cgi#set_header
    ~cache:`No_cache
    ~content_type:"text/javascript; charset=\"utf-8\""
    ();
  let target = Hashtbl.find arg_map "target" and
      ret = ref (Json_type.Array []) in
  ignore(Database.DB.query
           ("select date,content,url,author from comments where target=\"" ^ target^ "\"order by id desc")
           (comments_callback ret));
  Json_io.string_of_json !ret;;
    
let post_comment (cgi:Netcgi.cgi) arg_map () =
  cgi#set_header
    ~cache:`No_cache
    ~content_type:"text/javascript; charset=\"utf-8\""
    ();
  let target = cgi#argument_value "target" and
      author = cgi#argument_value "name" and
      email = cgi#argument_value "email" and
      url = cgi#argument_value "url" and
      content = cgi#argument_value "content" in
  ignore(Database.DB.update_stmt
           "insert into comments (target,author,email,url,content,date) values (?,?,?,?,?,datetime('now'))"
           [`TEXT target; `TEXT author; `TEXT email; `TEXT url; `TEXT content]);
  let ret = Json_type.Object [
    ("status",Json_type.String "ok");
    ("author",Json_type.String author);
    ("url",Json_type.String url);
    ("content",Json_type.String content);
    ("date",Json_type.String "seconds ago");
  ] in Json_io.string_of_json ret;;

File: service/comments.py
#!/usr/bin/python
#-*- coding: utf-8 -*-
# author : KDr2
#

import web
import cgi
import json
import base64
import datetime

import utils
import router
import email_sender

def format_comment(cmt):
    ret={}
    keys=['id','parent_id','author','url','content']
    for k in keys:
        v=cmt.get(k,"")
        if isinstance(v,(str,unicode)):v=cgi.escape(v)
        ret[k]=v
    #endfor
    date=cmt.get('date')
    ret['date']=date.strftime("%D %H:%m")
    return ret

#@router.route(r'^/(.*)')
class x(object):

    def GET(self,target):
        return "URL:"+target

    
@router.route(r'/comments/get/(.*)')
class CommentsGetter(object):

    def GET(self,target):
        data={"target":target}
        comments=utils.db.select('comments',data,where='target=$target')
        comments=map(format_comment,comments)
        return json.dumps(comments)


@router.route(r'/comments/post')
class CommentsPoster(object):

    def POST(self):
        #return web.notfound("Sorry")

        comments = web.input()
        if not hasattr(comments,'target') or \
                not hasattr(comments,'content') or\
                not hasattr(comments,'name'):
            return web.internalerror("Bad Post.")
        data={'date':web.SQLLiteral("NOW()")}
        #date=datetime.datetime.now()
        data['parent_id']=comments.get('parent_id',0)
        data['target']=comments.get('target','about.html')
        data['author']=comments.get('name','anonym')
        data['url']=comments.get('url','http://kdr2.net')
        data['email']=comments.get('email','')
        data['content']=comments.get('content','nothing')
        sequence_id = utils.db.insert('comments', **data)
        data['id']=sequence_id
        data['date']=datetime.datetime.now()
        try:
            r=base64.decodestring('a2lsbHkuZHJhd0BnbWFpbC5jb20=\n')
            subject=u'[KDr2.net]: New Comment from [%s]' % data['author']
            content=u'Comment Content:\n\n%s\n' % data['content']
            email_sender.sendmail_text(r,subject,content)
        except:
            pass
        return json.dumps([format_comment(data)])



    
